File: backend/pkg/database/repository/content_topic_match.go
package repository

import (
	"context"

	"backend/pkg/core"
	"backend/pkg/database/models"
	"backend/pkg/types"

	"github.com/google/uuid"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
)

// ContentTopicMatchRepository provides data access for content-topic matches
type ContentTopicMatchRepository struct {
	db *gorm.DB
}

// NewContentTopicMatchRepository creates a new ContentTopicMatchRepository
func NewContentTopicMatchRepository(db *gorm.DB) *ContentTopicMatchRepository {
	return &ContentTopicMatchRepository{db: db}
}

// contentTopicMatchModelToType converts a database model to types
func contentTopicMatchModelToType(m *models.ContentTopicMatch) *types.ContentTopicMatch {
	return &types.ContentTopicMatch{
		ID:              m.ID,
		ContentID:       m.ContentID,
		TopicID:         m.TopicID,
		SimilarityScore: m.SimilarityScore,
		IsPrimary:       m.IsPrimary,
		CreatedAt:       m.CreatedAt,
	}
}

// contentTopicMatchTypeToModel converts a types type to database model
func contentTopicMatchTypeToModel(m *types.ContentTopicMatch) *models.ContentTopicMatch {
	return &models.ContentTopicMatch{
		ID:              m.ID,
		ContentID:       m.ContentID,
		TopicID:         m.TopicID,
		SimilarityScore: m.SimilarityScore,
		IsPrimary:       m.IsPrimary,
		CreatedAt:       m.CreatedAt,
	}
}

// FindByContentID retrieves all topic matches for a content item
func (r *ContentTopicMatchRepository) FindByContentID(ctx context.Context, contentID uuid.UUID) ([]types.ContentTopicMatch, error) {
	var matchModels []models.ContentTopicMatch
	if err := r.db.WithContext(ctx).
		Where("content_id = ?", contentID).
		Order("similarity_score DESC").
		Find(&matchModels).Error; err != nil {
		return nil, err
	}

	matches := make([]types.ContentTopicMatch, len(matchModels))
	for i, m := range matchModels {
		matches[i] = *contentTopicMatchModelToType(&m)
	}
	return matches, nil
}

// FindByTopicID retrieves all content matches for a topic
func (r *ContentTopicMatchRepository) FindByTopicID(ctx context.Context, topicID uuid.UUID, offset, limit int) ([]types.ContentTopicMatch, int64, error) {
	var matchModels []models.ContentTopicMatch
	var total int64

	if err := r.db.WithContext(ctx).Model(&models.ContentTopicMatch{}).Where("topic_id = ?", topicID).Count(&total).Error; err != nil {
		return nil, 0, err
	}

	if err := r.db.WithContext(ctx).
		Where("topic_id = ?", topicID).
		Order("similarity_score DESC").
		Offset(offset).
		Limit(limit).
		Find(&matchModels).Error; err != nil {
		return nil, 0, err
	}

	matches := make([]types.ContentTopicMatch, len(matchModels))
	for i, m := range matchModels {
		matches[i] = *contentTopicMatchModelToType(&m)
	}
	return matches, total, nil
}

// FindPrimaryByTopicID retrieves all primary matches for a topic
func (r *ContentTopicMatchRepository) FindPrimaryByTopicID(ctx context.Context, topicID uuid.UUID, offset, limit int) ([]types.ContentTopicMatch, int64, error) {
	var matchModels []models.ContentTopicMatch
	var total int64

	if err := r.db.WithContext(ctx).Model(&models.ContentTopicMatch{}).
		Where("topic_id = ? AND is_primary = ?", topicID, true).
		Count(&total).Error; err != nil {
		return nil, 0, err
	}

	if err := r.db.WithContext(ctx).
		Where("topic_id = ? AND is_primary = ?", topicID, true).
		Order("similarity_score DESC").
		Offset(offset).
		Limit(limit).
		Find(&matchModels).Error; err != nil {
		return nil, 0, err
	}

	matches := make([]types.ContentTopicMatch, len(matchModels))
	for i, m := range matchModels {
		matches[i] = *contentTopicMatchModelToType(&m)
	}
	return matches, total, nil
}

// Save creates or updates a content-topic match (upserts)
func (r *ContentTopicMatchRepository) Save(ctx context.Context, match *types.ContentTopicMatch) error {
	model := contentTopicMatchTypeToModel(match)
	if model.ID == uuid.Nil {
		model.ID = uuid.New()
		match.ID = model.ID
	}

	return r.db.WithContext(ctx).Clauses(clause.OnConflict{
		Columns:   []clause.Column{{Name: "content_id"}, {Name: "topic_id"}},
		DoUpdates: clause.AssignmentColumns([]string{"similarity_score", "is_primary"}),
	}).Create(model).Error
}

// SaveBatch creates or updates multiple content-topic matches
func (r *ContentTopicMatchRepository) SaveBatch(ctx context.Context, matches []types.ContentTopicMatch) error {
	if len(matches) == 0 {
		return nil
	}

	matchModels := make([]models.ContentTopicMatch, len(matches))
	for i, m := range matches {
		if m.ID == uuid.Nil {
			matches[i].ID = uuid.New()
		}
		matchModels[i] = *contentTopicMatchTypeToModel(&matches[i])
	}

	return r.db.WithContext(ctx).Clauses(clause.OnConflict{
		Columns:   []clause.Column{{Name: "content_id"}, {Name: "topic_id"}},
		DoUpdates: clause.AssignmentColumns([]string{"similarity_score", "is_primary"}),
	}).Create(&matchModels).Error
}

// DeleteByContentID removes all topic matches for a content item
func (r *ContentTopicMatchRepository) DeleteByContentID(ctx context.Context, contentID uuid.UUID) error {
	return r.db.WithContext(ctx).Where("content_id = ?", contentID).Delete(&models.ContentTopicMatch{}).Error
}

// DeleteByTopicID removes all content matches for a topic
func (r *ContentTopicMatchRepository) DeleteByTopicID(ctx context.Context, topicID uuid.UUID) error {
	return r.db.WithContext(ctx).Where("topic_id = ?", topicID).Delete(&models.ContentTopicMatch{}).Error
}

// Delete removes a specific content-topic match
func (r *ContentTopicMatchRepository) Delete(ctx context.Context, id uuid.UUID) error {
	result := r.db.WithContext(ctx).Delete(&models.ContentTopicMatch{}, id)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return core.ErrNotFound
	}
	return nil
}

// CountByTopicID counts matches for a topic
func (r *ContentTopicMatchRepository) CountByTopicID(ctx context.Context, topicID uuid.UUID) (int64, error) {
	var count int64
	if err := r.db.WithContext(ctx).Model(&models.ContentTopicMatch{}).
		Where("topic_id = ?", topicID).
		Count(&count).Error; err != nil {
		return 0, err
	}
	return count, nil
}

// UpdatePrimaryStatus sets the is_primary flag for a match
func (r *ContentTopicMatchRepository) UpdatePrimaryStatus(ctx context.Context, contentID uuid.UUID, topicID uuid.UUID, isPrimary bool) error {
	return r.db.WithContext(ctx).Model(&models.ContentTopicMatch{}).
		Where("content_id = ? AND topic_id = ?", contentID, topicID).
		Update("is_primary", isPrimary).Error
}

// ClearPrimaryForContent clears primary flag for all matches of a content item
func (r *ContentTopicMatchRepository) ClearPrimaryForContent(ctx context.Context, contentID uuid.UUID) error {
	return r.db.WithContext(ctx).Model(&models.ContentTopicMatch{}).
		Where("content_id = ?", contentID).
		Update("is_primary", false).Error
}

File: backend/src/database/repository/content_topic_match.rs
use async_trait::async_trait;
use diesel::{ExpressionMethods, QueryDsl, SelectableHelper, dsl::count_star};
use diesel_async::RunQueryDsl;
use uuid::Uuid;

use crate::{
    core::insight::{ContentTopicMatch, ContentTopicMatchRepository},
    database::{
        models::content_topic_match::{ContentTopicMatchRow, NewContentTopicMatchRow},
        pool::PgPool,
    },
    error::AppError,
    schema::content_topic_match,
};

#[derive(Clone)]
pub struct DieselContentTopicMatchRepository {
    pool: PgPool,
}

impl DieselContentTopicMatchRepository {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    async fn connection(
        &self,
    ) -> Result<
        diesel_async::pooled_connection::deadpool::Object<diesel_async::AsyncPgConnection>,
        AppError,
    > {
        self.pool.get().await.map_err(|_| AppError::Database)
    }

    pub async fn find_by_content_id(
        &self,
        content_id: Uuid,
    ) -> Result<Vec<ContentTopicMatch>, AppError> {
        let mut connection = self.connection().await?;
        content_topic_match::table
            .filter(content_topic_match::content_id.eq(content_id))
            .order(content_topic_match::similarity_score.desc())
            .select(ContentTopicMatchRow::as_select())
            .load(&mut connection)
            .await
            .map(|rows: Vec<ContentTopicMatchRow>| rows.into_iter().map(Into::into).collect())
            .map_err(|_| AppError::Database)
    }

    pub async fn find_by_topic_id(
        &self,
        topic_id: Uuid,
        offset: i64,
        limit: i64,
    ) -> Result<(Vec<ContentTopicMatch>, i64), AppError> {
        self.find_by_topic_id_with_primary(topic_id, offset, limit, None)
            .await
    }

    pub async fn save(&self, value: &mut ContentTopicMatch) -> Result<(), AppError> {
        self.save_batch(std::slice::from_mut(value)).await
    }

    pub async fn delete_by_content_id(&self, content_id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        diesel::delete(
            content_topic_match::table.filter(content_topic_match::content_id.eq(content_id)),
        )
        .execute(&mut connection)
        .await
        .map(|_| ())
        .map_err(|_| AppError::Database)
    }

    pub async fn delete_by_topic_id(&self, topic_id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        diesel::delete(
            content_topic_match::table.filter(content_topic_match::topic_id.eq(topic_id)),
        )
        .execute(&mut connection)
        .await
        .map(|_| ())
        .map_err(|_| AppError::Database)
    }

    pub async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        match diesel::delete(content_topic_match::table.find(id))
            .execute(&mut connection)
            .await
            .map_err(|_| AppError::Database)?
        {
            0 => Err(AppError::NotFound),
            _ => Ok(()),
        }
    }

    pub async fn update_primary_status(
        &self,
        content_id: Uuid,
        topic_id: Uuid,
        is_primary: bool,
    ) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        diesel::update(
            content_topic_match::table
                .filter(content_topic_match::content_id.eq(content_id))
                .filter(content_topic_match::topic_id.eq(topic_id)),
        )
        .set(content_topic_match::is_primary.eq(is_primary))
        .execute(&mut connection)
        .await
        .map(|_| ())
        .map_err(|_| AppError::Database)
    }

    pub async fn clear_primary_for_content(&self, content_id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        diesel::update(
            content_topic_match::table.filter(content_topic_match::content_id.eq(content_id)),
        )
        .set(content_topic_match::is_primary.eq(false))
        .execute(&mut connection)
        .await
        .map(|_| ())
        .map_err(|_| AppError::Database)
    }

    async fn find_by_topic_id_with_primary(
        &self,
        topic_id: Uuid,
        offset: i64,
        limit: i64,
        primary: Option<bool>,
    ) -> Result<(Vec<ContentTopicMatch>, i64), AppError> {
        let mut connection = self.connection().await?;
        let mut count_query = content_topic_match::table
            .filter(content_topic_match::topic_id.eq(topic_id))
            .into_boxed();
        if let Some(primary) = primary {
            count_query = count_query.filter(content_topic_match::is_primary.eq(primary));
        }
        let total = count_query
            .select(count_star())
            .first(&mut connection)
            .await
            .map_err(|_| AppError::Database)?;

        let mut query = content_topic_match::table
            .filter(content_topic_match::topic_id.eq(topic_id))
            .order(content_topic_match::similarity_score.desc())
            .offset(offset)
            .into_boxed();
        if let Some(primary) = primary {
            query = query.filter(content_topic_match::is_primary.eq(primary));
        }
        if limit >= 0 {
            query = query.limit(limit);
        }
        let rows = query
            .select(ContentTopicMatchRow::as_select())
            .load(&mut connection)
            .await
            .map_err(|_| AppError::Database)?;
        Ok((rows.into_iter().map(Into::into).collect(), total))
    }
}

#[async_trait]
impl ContentTopicMatchRepository for DieselContentTopicMatchRepository {
    async fn save_batch(&self, matches: &mut [ContentTopicMatch]) -> Result<(), AppError> {
        if matches.is_empty() {
            return Ok(());
        }
        let now = chrono::Utc::now();
        for value in matches.iter_mut() {
            if value.id.is_nil() {
                value.id = Uuid::new_v4();
            }
        }
        let rows: Vec<_> = matches
            .iter()
            .map(|value| NewContentTopicMatchRow {
                id: value.id,
                content_id: value.content_id,
                topic_id: value.topic_id,
                similarity_score: value.similarity_score,
                is_primary: value.is_primary,
                created_at: value.created_at.unwrap_or(now),
            })
            .collect();
        let mut connection = self.connection().await?;
        diesel::insert_into(content_topic_match::table)
            .values(rows)
            .on_conflict((
                content_topic_match::content_id,
                content_topic_match::topic_id,
            ))
            .do_update()
            .set((
                content_topic_match::similarity_score.eq(diesel::upsert::excluded(
                    content_topic_match::similarity_score,
                )),
                content_topic_match::is_primary
                    .eq(diesel::upsert::excluded(content_topic_match::is_primary)),
            ))
            .execute(&mut connection)
            .await
            .map(|_| ())
            .map_err(|_| AppError::Database)
    }

    async fn count_by_topic_id(&self, topic_id: Uuid) -> Result<i64, AppError> {
        let mut connection = self.connection().await?;
        content_topic_match::table
            .filter(content_topic_match::topic_id.eq(topic_id))
            .select(count_star())
            .first(&mut connection)
            .await
            .map_err(|_| AppError::Database)
    }

    async fn find_primary_by_topic_id(
        &self,
        topic_id: Uuid,
        offset: i64,
        limit: i64,
    ) -> Result<(Vec<ContentTopicMatch>, i64), AppError> {
        self.find_by_topic_id_with_primary(topic_id, offset, limit, Some(true))
            .await
    }
}

impl From<ContentTopicMatchRow> for ContentTopicMatch {
    fn from(row: ContentTopicMatchRow) -> Self {
        Self {
            id: row.id,
            content_id: row.content_id,
            topic_id: row.topic_id,
            similarity_score: row.similarity_score,
            is_primary: row.is_primary.unwrap_or_default(),
            created_at: row.created_at,
        }
    }
}
